python/shotgun_model/shotgun_entity_model.py

```diff
@@ -0,0 +1,249 @@
+
+        # load the data from the cache:
+        self._load_data(entity_type, filters, hierarchy, fields)
+
+    def destroy(self):
+        """
+        Call to clean-up the model when it is finished with
+        """
+        ShotgunModel.destroy(self)
+        self._entity_icons = {}
+        self._step_swatch_icons = {}
+        self._default_icon = None
+
+    def get_entity_icon(self, entity_type):
+        """
+        Retrieve the icon for the specified entity type if available.
+
+        :param entity_type: The entity type to retrieve the icon for
+        :returns:           A QIcon if an icon was found for the specified entity
+                            type, otherwise None.
+        """
+        icon = None
+        if entity_type in self._entity_icons:
+            # we've previously asked for the icon
+            icon = self._entity_icons[entity_type]
+        else:
+            # see if we have the icon in the resources:
+            icon_path = ":/tk-framework-shotgunutils/icon_%s_dark.png" % entity_type
+            if QtCore.QFile.exists(icon_path): 
+                # create the new icon from this resource:
+                icon = QtGui.QIcon(QtGui.QPixmap(icon_path))
+            self._entity_icons[entity_type] = icon
+
+        return QtGui.QIcon(icon) if icon else None
+
+    def get_entities(self, item):
+        """
+        Get entities for the current item by traversing up the tree and pulling 
+        entity information from each item if possible
+
+        :param item:    The item to find entities for
+        :returns:       A list of entities in the order they were found starting
+                        from the specified item.
+        """
+        current_item = item
+        entities = []
+        while current_item:
+            item_entity = self.get_entity(current_item)
+            if item_entity:
+                entities.append(item_entity)
+            current_item = current_item.parent()
+
+        return entities
+
+    def get_entity(self, item):
+        """
+        Get the Shotgun entity details for the specified model item.
+
+        :param item:    The item to retrieve the entity details for
+        :returns:       A Shotgun entity dictionary for the item if it represents
+                        and entity, otherwise None
+        """
+        # first, if this is a leaf item then it will represent an entity:
+        sg_data = item.get_sg_data()
+        if sg_data:
+            return sg_data
+
+        # item doesn't represent an entity directly so look for an entity in the field data instead:
+        field_data = get_sanitized_data(item, self.SG_ASSOCIATED_FIELD_ROLE)
+        field_value = field_data.get("value")
+        if (field_value 
+            and isinstance(field_value, dict) 
+            and "id" in field_value 
+            and "type" in field_value):
+            return field_value
+
+        return None
+
+    def async_refresh(self):
```

Comment:
I'm going to leave it as is for the merge as it's consistent with everywhere else but I agree - refresh() would work fine